### PR TITLE
KDS implementation for sw emulation

### PR DESCRIPTION
**Pull request by vboggara-xilinx**
Implemented KDS for SW emulation. this code is under param control .ini file. 
Runtime.kds_sw_emu is the param to enable the kds for sw emulation.

### CODE REVIEW

File: src/runtime_src/core/common/xclbin_parser.cpp [hunk 1]
```diff
@@ -387,5 +391,52 @@ get_kernel_freq(const axlf* top)
...
+    auto begin = reinterpret_cast<const char*>(top) + xml_hdr->m_sectionOffset;
+    const char *xml_data = reinterpret_cast<const char*>(begin);
+    uint64_t xml_size = xml_hdr->m_sectionSize;
+
+    pt::ptree xml_project;
+    std::stringstream xml_stream;
+    xml_stream.write(xml_data, xml_size);
+    pt::read_xml(xml_stream, xml_project);
+
+    auto xml_plotform = xml_project.get_child_optional("project.platform");
```
Comment (stsoe) [inline src/runtime_src/core/common/xclbin_parser.cpp]: Spelling (platform)

File: src/runtime_src/core/common/xclbin_parser.cpp [hunk 2]
```diff
@@ -387,5 +391,52 @@ get_kernel_freq(const axlf* top)
...
+                  continue;
+                auto base = convert(xml_remap.second.get<std::string>("<xmlattr>.base"));
+                addrVec.push_back(base);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
```
Comment (stsoe) [inline src/runtime_src/core/common/xclbin_parser.cpp]: Too many levels of nesting. Please restructure with early exits.
For example, first two levels of indent can be eliminated with
```
if (!xml_hdr)
  return addrVec;
...
if (!xml_platform)
 return addrVec;
```
Any indents after that can be left as is, though factoring out into separate embedded lambda functions may be cleaner; don't worry about this though.

File: src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx [hunk 3]
```diff
@@ -968,8 +975,18 @@ namespace xclcpuemhal2 {
...
       mLogStream << __func__ << ", " << std::this_thread::get_id() << std::endl;
     }
-    if(!sock)
+    if (!sock) {
+      PRINTENDFUNC
+      if (mIsKdsSwEmu && mSWSch && mCore)
+      {
+        mSWSch->fini_scheduler_thread();
+        delete mCore;
+        mCore = NULL;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx]: in C++ NULL should be replaced by nullptr

File: src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx [hunk 4]
```diff
@@ -1035,6 +1069,14 @@ namespace xclcpuemhal2 {
   
   CpuemShim::~CpuemShim() 
   {
+    if (mIsKdsSwEmu && mSWSch && mCore)
+    {
+      mSWSch->fini_scheduler_thread();
+      delete mCore;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx]: deletes look suspicious.  should this heap variables be managed in a smart pointer?

File: src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/swscheduler.h [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    int		trigger;
+    std::mutex mLock;
+  };
+
+  class xocl_sched
+  {
+    public:
+      pthread_t                   scheduler_thread;
+      pthread_mutex_t             state_lock;
+      pthread_cond_t              state_cond;
```
Comment (stsoe) [inline src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/swscheduler.h]: No, please use std::thread library.  Do *not* use pthread APIs directly.

File: src/runtime_src/xocl/core/device.cpp [hunk 6]
```diff
@@ -221,7 +221,8 @@ init_scheduler(xocl::device* device)
     throw xocl::error(CL_INVALID_PROGRAM,"Cannot initialize MBS before program is loadded");
 
   auto axlf = device->get_axlf();
-  if (is_sw_emulation()) {
+  bool isKdsSwEmu = xrt_core::config::get_flag_kds_sw_emu();
```
Comment (stsoe) [inline src/runtime_src/xocl/core/device.cpp]: Follow naming convention in file.  There are no camel case variable names in this file.

File: src/runtime_src/xocl/core/device.cpp [hunk 7]
```diff
@@ -221,7 +221,8 @@ init_scheduler(xocl::device* device)
...
+  if (is_sw_emulation() && !isKdsSwEmu) {
```
Comment (stsoe) [inline src/runtime_src/xocl/core/device.cpp]: Remove the iskdsswemu variable and fold into if condition.

File: src/runtime_src/xrt/scheduler/scheduler.cpp [hunk 8]
```diff
@@ -53,8 +53,9 @@ is_sw_emulation()
 
 inline bool
 kds_enabled(bool forceoff=false)
-{
-  static bool enabled = !is_sw_emulation() && xrt::config::get_kds() && !xrt::config::get_feature_toggle("Runtime.sws") && !is_windows();
+{  
+  bool isKdsEmu = is_sw_emulation() ? (xrt_core::config::get_flag_kds_sw_emu() ? false : true) : false;
```
Comment (stsoe) [inline src/runtime_src/xrt/scheduler/scheduler.cpp]: No camel case please.

File: src/runtime_src/xrt/scheduler/sws.cpp [hunk 9]
```diff
@@ -857,6 +857,8 @@ start()
   if (s_running)
     throw std::runtime_error("software command scheduler is already started");
 
+  std::cout << "SWS Thread started " << std::endl;
```
Comment (stsoe) [inline src/runtime_src/xrt/scheduler/sws.cpp]: Remove spurious print.

**Comment by salindac** [comment]
retest this please

File: src/runtime_src/core/common/xclbin_parser.cpp (hunk 1 above)
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/xclbin_parser.cpp]: changed to platform

File: src/runtime_src/core/common/xclbin_parser.cpp (hunk 2 above)
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/xclbin_parser.cpp]: addressed this comment in the code.

File: src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx (hunk 3 above)
Comment (vboggara-xilinx) [reply src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx]: changed to nullptr

File: src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx (hunk 4 above)
Comment (vboggara-xilinx) [reply src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/shim.cxx]: this line gets called during end of program execution.

File: src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/swscheduler.h (hunk 5 above)
Comment (vboggara-xilinx) [reply src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/swscheduler.h]: changed to std::thread library

File: src/runtime_src/xocl/core/device.cpp (hunk 6 above)
Comment (vboggara-xilinx) [reply src/runtime_src/xocl/core/device.cpp]: removed isKdsSwEmu variable

File: src/runtime_src/xocl/core/device.cpp (hunk 7 above)
Comment (vboggara-xilinx) [reply src/runtime_src/xocl/core/device.cpp]: added xrt_core::config::get_flag_kds_sw_emu() to if condition

File: src/runtime_src/xrt/scheduler/scheduler.cpp (hunk 8 above)
Comment (vboggara-xilinx) [reply src/runtime_src/xrt/scheduler/scheduler.cpp]: removed camel case

File: src/runtime_src/xrt/scheduler/sws.cpp (hunk 9 above)
Comment (vboggara-xilinx) [reply src/runtime_src/xrt/scheduler/sws.cpp]: Moved couts to pragma scope

File: src/runtime_src/xrt/scheduler/sws.cpp [hunk 10]
```diff
@@ -856,7 +858,9 @@ start()
 {
   if (s_running)
     throw std::runtime_error("software command scheduler is already started");
-
+#ifdef EM_DEBUG_SWS
```
Comment (stsoe) [inline src/runtime_src/xrt/scheduler/sws.cpp]: Please remove or use the XRT_DEBUG infra structure. Let's not pollute code with ifdefs.

Comment (vboggara-xilinx) [reply src/runtime_src/xrt/scheduler/sws.cpp]: used XRT_DEBUG infrastructure

File: src/runtime_src/core/common/xclbin_parser.cpp [hunk 11]
```diff
@@ -387,5 +391,57 @@ get_kernel_freq(const axlf* top)
   return kernel_clk_freq;
 }
 
+
+std::vector<uint64_t>
+get_kernel_inst_addrs(const axlf* top)
+{
+  std::vector<uint64_t> addrVec;
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/common/xclbin_parser.cpp]: Can you please add a comment for this code saying that ,this code is going to be removed once rtd file is generated for sw_emu.

Comment (chvamshi-xilinx) [reply src/runtime_src/core/common/xclbin_parser.cpp]: We are planning to remove the EMBEDDED_METADATA section usage in 2020.1+

Comment (vboggara-xilinx) [reply src/runtime_src/core/common/xclbin_parser.cpp]: comment added in xclbin_parser.cpp

**Comment by vboggara-xilinx** [comment]
since pull request got approved, can someone merge these changes to master branch?

**Comment by stsoe** [comment]
Please address the conflicts, then we can merge after successful build.

**Comment by vboggara-xilinx** [comment]
Resolved merge conflicts and pushed the changes, please merge these changes to master branch,

**Comment by chvamshi-xilinx** [comment]
retest this please

**Pull request merged**